File: client-android/app/src/androidTest/java/stasis/client_android/mocks/MockAnalyticsPersistence.kt
package stasis.client_android.mocks

import stasis.client_android.lib.telemetry.analytics.AnalyticsEntry
import stasis.client_android.lib.telemetry.analytics.AnalyticsPersistence
import stasis.client_android.lib.utils.Try
import java.time.Instant
import java.util.concurrent.atomic.AtomicReference

open class MockAnalyticsPersistence(private val existing: Try<AnalyticsEntry?>) : AnalyticsPersistence {
    private val cachedEntries: MutableList<AnalyticsEntry> = mutableListOf()
    private val transmittedEntries: MutableList<AnalyticsEntry> = mutableListOf()

    private val lastCachedRef: AtomicReference<Instant> = AtomicReference(Instant.MIN)
    private val lastTransmittedRef: AtomicReference<Instant> = AtomicReference(Instant.MIN)

    override fun cache(entry: AnalyticsEntry) {
        lastCachedRef.set(Instant.now())
        cachedEntries.add(entry)
    }

    override suspend fun transmit(entry: AnalyticsEntry): Try<Unit> {
        lastTransmittedRef.set(Instant.now())
        transmittedEntries.add(entry)
        return Try.Success(Unit)
    }

    override suspend fun restore(): Try<AnalyticsEntry?> =
        existing

    override val lastCached: Instant
        get() = lastCachedRef.get()

    override val lastTransmitted: Instant
        get() = lastTransmittedRef.get()

    val cached: List<AnalyticsEntry>
        get() = cachedEntries

    val transmitted: List<AnalyticsEntry>
        get() = transmittedEntries
}

File: layers/src/test/scala/stasis/layers/telemetry/analytics/MockAnalyticsPersistence.scala
package stasis.layers.telemetry.analytics

import java.time.Instant
import java.util.concurrent.atomic.AtomicReference

import scala.collection.mutable
import scala.concurrent.Future
import scala.util.Failure
import scala.util.Success
import scala.util.Try

import org.apache.pekko.Done

class MockAnalyticsPersistence(existing: Try[Option[AnalyticsEntry]]) extends AnalyticsPersistence {
  override type Self = this.type

  private val cachedEntries: mutable.Queue[AnalyticsEntry] = mutable.Queue.empty
  private val transmittedEntries: mutable.Queue[AnalyticsEntry] = mutable.Queue.empty

  private val lastCachedRef: AtomicReference[Instant] = new AtomicReference(Instant.EPOCH)
  private val lastTransmittedRef: AtomicReference[Instant] = new AtomicReference(Instant.EPOCH)

  override def cache(entry: AnalyticsEntry): Unit = {
    lastCachedRef.set(Instant.now())
    cachedEntries.enqueue(entry)
  }

  override def transmit(entry: AnalyticsEntry): Future[Done] = {
    lastTransmittedRef.set(Instant.now())
    transmittedEntries.enqueue(entry)
    Future.successful(Done)
  }

  override def restore(): Future[Option[AnalyticsEntry]] =
    Future.fromTry(existing)

  override def lastCached: Instant = lastCachedRef.get()

  override def lastTransmitted: Instant = lastTransmittedRef.get()

  def cached: Seq[AnalyticsEntry] = cachedEntries.toSeq

  def transmitted: Seq[AnalyticsEntry] = transmittedEntries.toSeq

  override def withClientProvider(provider: AnalyticsClient.Provider): Self =
    this
}

object MockAnalyticsPersistence {
  def apply(): MockAnalyticsPersistence =
    new MockAnalyticsPersistence(existing = Success(None))

  def apply(existing: AnalyticsEntry): MockAnalyticsPersistence =
    new MockAnalyticsPersistence(existing = Success(Some(existing)))

  def apply(existing: Throwable): MockAnalyticsPersistence =
    new MockAnalyticsPersistence(existing = Failure(existing))
}
